rules/prefer-error-cause.js

```diff
@@ -0,0 +1,364 @@
+
+	switch (statementToFix.type) {
+		case 'VariableDeclarator': {
+			nodeToInsertArgument = statementToFix.init;
+			break;
+		}
+
+		case 'AssignmentExpression': {
+			nodeToInsertArgument = statementToFix.right;
+			break;
+		}
+
+		case 'ThrowStatement': {
+			nodeToInsertArgument = statementToFix.argument;
+			break;
+		}
+
+		default: {
+			throw new Error('statementToFix is not invalid type');
+		}
+	}
+
+	return fixerUtils.insertCauseArgument({
+		fixer,
+		node: nodeToInsertArgument,
+		sourceCode,
+		value: errorArgumentIdentifier,
+	});
+};
+
+const getAllNodesToFix = ({node, context, throwStatement}) => {
+	const result = [];
+
+	let statementToFix;
+	let errorConstructorLastArgument;
+
+	// `try {} catch (err) { throw new Error('oops', {cause: err})}`
+	// `promise.catch(err => { throw new Error('oops', {cause: err})`
+	if (throwStatement.argument.type === 'NewExpression') {
+		errorConstructorLastArgument = peekTop(throwStatement.argument.arguments);
+
+		if (!errorConstructorLastArgument) {
+			reportCannotBeFixed(node, context);
+			return;
+		}
+
+		result.push({
+			node: errorConstructorLastArgument,
+			statementToFix: throwStatement,
+		});
+	} else {
+		// Assume Error's constructor exists in other node of the block
+		// It could be VariableDeclarator or AssignmentExpression.
+
+		const newExpressionSelector = reference =>
+			(reference.identifier?.parent.type === 'AssignmentExpression'
+			|| reference.identifier?.parent.type === 'VariableDeclarator')
+			&& reference.writeExpr?.type === 'NewExpression'
+			&& reference.resolved?.name === throwStatement.argument.name;
+
+		const thrownErrorDeclarators = getReferences(context.getScope())
+			.filter(reference => newExpressionSelector(reference))
+			.map(reference => reference.identifier.parent);
+
+		// Report newExpression not found in given block.
+		// `try {} catch { const err = new Error; try {} catch { throw err; } }`
+		if (_.isEmpty(thrownErrorDeclarators)) {
+			reportCannotBeFixed(node, context);
+			return;
+		}
+
+		// `try {} catch (err) { let e1 = new Error('oops'); e1 = new Error('oops'); throw error;}`
+
+		for (const thrownErrorDeclarator of thrownErrorDeclarators) {
+			statementToFix = thrownErrorDeclarator;
+
+			let targetArguments;
+			if (thrownErrorDeclarator.type === 'VariableDeclarator') {
+				targetArguments = thrownErrorDeclarator.init.arguments;
+			} else if (thrownErrorDeclarator.type === 'AssignmentExpression') {
+				targetArguments = thrownErrorDeclarator.right.arguments;
+			}
+
+			errorConstructorLastArgument = peekTop(targetArguments);
+
+			if (!errorConstructorLastArgument) {
+				reportCannotBeFixed(node, context);
+				return;
+			}
+
+			result.push({
+				node: errorConstructorLastArgument,
+				statementToFix,
+			});
+		}
+	}
+
+	return result;
+};
+
+const handleCatchBlock = ({node, context, statements, parameter}) => {
+	const throwStatement = statements.find(statement => statement.type === 'ThrowStatement');
```

Comment:
This is not right

```js
try {} catch (error) {
	if (error instanceof SynctaxError) {
		return ''
	} else {
		throw new Error('')
	}
}
```

Reply:
Feel hard to implement this, we should make sure the `throw` is in correct scope. Maybe check ancestors to get `function` and `try/catch`, make sure they are in current one?

Reply:
> Feel hard to implement this, we should make sure the `throw` is in correct scope. Maybe check ancestors to get `function` and `try/catch`, make sure they are in current one?

Thanks for pointing this out.
IMHO, we might need to use `esquery` here because throwStatement could emerge in any childnode.
If using esquery is too costly, how about trying to check closest ancestor using `getMatchingAncestorOfType` of `prefer-keyboard-event.key.js`?
I just updated the code using `esquery` for trying to correct this, if this is still incorrect or not in the right direction, please let me know.

Reply:
That should work, there is another similar on https://github.com/sindresorhus/eslint-plugin-unicorn/blob/c1cb41a67a5449e58bae60ceeb97279482b60354/rules/no-array-for-each.js#L49

Reply:
> we might need to use esquery here because throwStatement could emerge in any childnode.

Maybe we don't need, we can listen on throw expression and find ancestor we want(try-catch or promise callback). Instead of get the ancestor first.

Reply:
For example, in this case https://github.com/sindresorhus/eslint-plugin-unicorn/pull/1793#discussion_r860563200, we don't need care about the outer one at all.

Reply:
> Maybe we don't need, we can listen on throw expression and find ancestor we want(try-catch or promise callback). Instead of get the ancestor first.

I just updated the code by changing the selector to just `ThrowStatement`, and using `getMatchingAncestorOfType` to find out catchBlock.